### PR TITLE
refactoring: admin / static pages / fetch data / redux

### PR DESCRIPTION
**Pull request by Drapegnik**
## done
- remove auth / login form / formix
- remove all admin components
- enable [static page generation (SSG)](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation)
- fetch data via [`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering)
- remove `redux`
- remove legacy code
- replace [qs](https://bundlephobia.com/result?p=qs@6.9.4) -> [querystring](https://bundlephobia.com/result?p=querystring@0.2.0)
- load `ImageSlider` via [dynamic import](https://nextjs.org/docs/advanced-features/dynamic-import)

### before
```js
Page                                                           Size     First Load JS
┌ λ /                                                          2.82 kB         169 kB
├   /_app                                                      0 B             158 kB
├ λ /404                                                       2.97 kB         161 kB
├ λ /about                                                     5.44 kB         163 kB
├   └ css/6e65d8528558a09508ca.css                             1.12 kB
├ λ /admin/articles                                            4.77 kB         162 kB
├   └ css/77fee145c6aab19dd296.css                             1.15 kB
├ λ /admin/login                                               1.42 kB         159 kB
├   └ css/a22c22924ab9c8a81d38.css                             787 B
├ λ /admin/preview                                             5.26 kB         215 kB
├   └ css/a6f203003599ad674752.css                             2.91 kB
├ λ /article                                                   5.06 kB         214 kB
├   └ css/eb2fa2d1a42c38899ebb.css                             3.14 kB
├ λ /collection                                                2.17 kB         163 kB
├   └ css/0d727aad8902d410ef25.css                             1.44 kB
├ λ /diary                                                     2.21 kB         206 kB
├   └ css/85d2c6510216acd9e73d.css                             2.76 kB
├ λ /markup/example                                            869 B           158 kB
├   └ css/c015e6a81b30a0403463.css                             114 B
├ λ /status                                                    2.93 kB         161 kB
├ λ /tag                                                       2.47 kB         164 kB
├   └ css/b963d7b47f2a37c7088f.css                             1.28 kB
└ λ /topic                                                     1.54 kB         163 kB
    └ css/bdfafbbebd0a9b2121f2.css                             961 B
+ First Load JS shared by all                                  158 kB
  ├ chunks/12c19299e66099f9603e82337647d55a9c45f3e1.137439.js  19.9 kB
  ├ chunks/2908ed6f3bf86d7acc73f257b182787faef3c1f6.d294ce.js  14.3 kB
  ├ chunks/298356a432d016ec57c91bb61de8c3be37a68813.df08ae.js  3.95 kB
  ├ chunks/5b50eaef33aa706fd9baec48a31b48416096b606.7ac0c7.js  7.68 kB
  ├ chunks/89d6c0bf7c8752cf2e66e56e076eb225605c5ab2.0c65b9.js  14.7 kB
  ├ chunks/9993898066fb3a43ef984d6b1a07d90751c488f9.4e1b68.js  11.9 kB
  ├ chunks/cc42a434d90cb7f506008c6caa87e0d388fd79b9.97b485.js  3.22 kB
  ├ chunks/e56e55be1371db940227cd932da9955c834ca06e.1348c7.js  12.8 kB
  ├ chunks/framework.b11cd6.js                                 40.2 kB
  ├ chunks/main.0a8824.js                                      6.54 kB
  ├ chunks/pages/_app.70e0ab.js                                21.7 kB
  ├ chunks/webpack.e06743.js                                   751 B
  └ css/e1d81e68f72583ecbc3e.css                               21.6 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
```

### after
```js
Page                                                           Size     First Load JS
┌ ● / (ISR: 1 Seconds)                                         377 B           137 kB
├   /_app                                                      0 B             114 kB
├ λ /404                                                       2.21 kB         116 kB
├ ○ /about                                                     4.44 kB         118 kB
├   └ css/4b99dc700b5ed115c020.css                             832 B
├ λ /article                                                   3.68 kB         147 kB
├   └ css/d8d5fd40f8d2008bdca1.css                             2.96 kB
├ λ /collection                                                2.56 kB         122 kB
├   └ css/6575df9d97e6f00cebe1.css                             1.18 kB
├ λ /diary                                                     1.75 kB         139 kB
├   └ css/466469d3d2d9ccb5b7fc.css                             2.36 kB
├ ○ /status                                                    2.16 kB         116 kB
├ λ /tag                                                       2.41 kB         122 kB
├   └ css/af7554a004565612ae67.css                             1 kB
└ λ /topic                                                     1.56 kB         115 kB
    └ css/820f5ba2bfec52e744c0.css                             459 B
+ First Load JS shared by all                                  114 kB
  ├ chunks/59b7d1a0d58e1bc85160ee7c85fb167b0df65d10.2a81fc.js  3.94 kB
  ├ chunks/71b102af5771852fd65e83f704b16fc52311c10e.8fa717.js  7.45 kB
  ├ chunks/ce1ef5918f6b6f55be619a57dd625f25144c716e.719c8e.js  10.6 kB
  ├ chunks/commons.29f157.js                                   26.7 kB
  ├ chunks/framework.0d373e.js                                 40.4 kB
  ├ chunks/main.94fab3.js                                      6.54 kB
  ├ chunks/pages/_app.ac84c9.js                                16.7 kB
  ├ chunks/webpack.bdbd09.js                                   1.55 kB
  ├ css/1d902f9543b2bccb7481.css                               21.4 kB
  └ css/5ab46d48a911fb161a20.css                               399 B
```

## results

- First Load JS shared by all: `158 kB` -> `114 kB` = `-44 kB` (`28 %`)
- Main Page: `2.82 kB` -> `377 B`

### CODE REVIEW

**Comment by Drapegnik** [comment]
[next 9.3 static pages](https://trello.com/c/dLyBbyLq/494-next-93-static-pages)

**Comment by Drapegnik** [comment]
[refactor frontend: replace `redux` with hooks :)](https://trello.com/c/LN2WKroB/381-refactor-frontend-replace-redux-with-hooks)

**Comment by Drapegnik** [comment]
[remove legacy code](https://trello.com/c/ZnLY5ovF/504-remove-legacy-code)

**Comment by Drapegnik** [comment]
[remove legacy code](https://trello.com/c/ZnLY5ovF/504-remove-legacy-code)

File: components/articles/blocks/DiaryBlock.js [hunk 1]
```diff
@@ -1,85 +1,92 @@
...
+import { getDiary, isNextAvailable } from 'utils/features/diary';
+import api from 'constants/api';
 
-import { SHORT_DATE_FORMAT } from 'constants';
 import { DIARY_PICTURE_WIDTH } from 'constants/misc';
 import { ROUTES_NAMES } from 'routes';
-import styles from './diary.module.scss';
 
 import BlockWrapper from './BlockWrapper';
+import styles from './diary.module.scss';
```
Comment (UladBohdan) [inline components/articles/blocks/DiaryBlock.js]: how about always importing `styles` on the first line of the file?

File: components/common/layout/footer/SubscribeForm.js [hunk 2]
```diff
@@ -1,23 +1,59 @@
...
-import { homeActions } from 'redux/ducks/home';
-import { validEmail } from 'utils/validators';
-import createConstants from 'lib/utils/createConstants';
 import styles from './index.module.scss';
 
 const b = bem(styles);
-const STATUS = createConstants('subscribed', 'unsubscribed');
 
+const emailRegexp = /^(([^<>()[\].,;:\s@"]+(\.[^<>()[\].,;:\s@"]+)*)|(".+"))@(([^<>()[\].,;:\s@"]+\.)+[^<>()[\].,;:\s@"]{2,})$/;
+const isEmail = email => !!email?.match(emailRegexp);
```
Comment (UladBohdan) [inline components/common/layout/footer/SubscribeForm.js]: Move this to utils ? Or probably to `lib` repo

File: constants/api.js [hunk 3]
```diff
@@ -1,35 +1,20 @@
...
-    login: `${AUTH_URL}/login`,
-    register: `${AUTH_URL}/register`,
-    logout: `${AUTH_URL}/logout`,
-  },
   storage: {
     getMainPage: `${API_URL}/storage/main-page`,
     getSidebar: `${API_URL}/storage/sidebar`,
   },
   core: {
-    uploads: `${API_URL}/core/uploads`,
```
Comment (UladBohdan) [inline constants/api.js:19]: did it work before? :)

File: jsconfig.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "compilerOptions": {
+        "target": "esnext",
+        "module": "esnext",
+        "allowSyntheticDefaultImports": true,
+        "baseUrl": "./",
+    }
+}
```
Comment (UladBohdan) [inline jsconfig.json]: add a newline here

File: next.config.js [hunk 5]
```diff
@@ -1,8 +1,23 @@
...
+const getBundleAnalyzer = require('@next/bundle-analyzer');
+
 const ENV = require('./utils/env');
 
-module.exports = {
+const nextConfig = {
   env: {
     isProd: ENV === 'production',
     isStaging: ENV === 'staging',
+    isDev: ENV === 'development',
```
Comment (UladBohdan) [inline next.config.js]: probably `isDev: ENV != 'production' && ENV != 'staging'` ?

File: pages/admin/preview.js [hunk 6]
```diff
@@ -1,76 +0,0 @@
...
-import useBoolean from 'hooks/useBoolean';
-
-const b = bem(styles);
-
-const mapStateToProps = (state, { lang }) => ({
-  article: adminArticlesSelectors.getPreview(state, lang),
-  error: adminArticlesSelectors.getError(state),
-});
-
-const ArticlePreview = ({ article, error }) => {
```
Comment (UladBohdan) [inline pages/admin/preview.js:25]: I guess we should temporary disable preview-link in fibery then

btw what about preview mode? will it become a part of admin panel and live on admin subdomain?

File: pages/diary.js [hunk 7]
```diff
@@ -21,62 +19,39 @@ import {
...
+    slug,
+    author: { name, diaryImage: image },
+    date,
+    text: { content },
+  },
+  articles,
+  metaTitle,
+  metaKeywords,
+  shortContent,
+  basicText,
```
Comment (UladBohdan) [inline pages/diary.js]: `basicShareText` and `extendedShareText`

File: pages/diary.js [hunk 8]
```diff
@@ -21,62 +19,39 @@ import {
...
-import { diaryActions, diarySelectors } from 'redux/ducks/diary';
-import { formatLocalizedDate, getYear, isSameDate } from 'utils/formatters';
+import { formatLocalizedDate, DATE_FORMAT, SHORT_DATE_FORMAT } from 'utils/formatters/date';
 import fiberyRenderer from 'utils/fibery/renderer';
 import fiberyToString from 'utils/fibery/toString';
-import { populateRequest, makeRequest } from 'utils/request';
-import { DiaryShape, LangType } from 'utils/customPropTypes';
+import { makeRequest } from 'utils/request';
 import { getLocalizedArticles } from 'utils/getters';
+import { getDiary, getShareText, isNextAvailable } from 'utils/features/diary';
```
Comment (UladBohdan) [inline pages/diary.js]: `getDiaryShareText`, `isNextDiaryAvailable`

File: constants/api.js (hunk 3 above)
Comment (Drapegnik) [reply constants/api.js:19]: should upload to cloudinary:
https://github.com/babajka/babajka-backend/blob/dfd78f437999d7f4fc0078a7a2eb1ed5c9fe1c40/src/api/core/controller.js#L12

File: pages/admin/preview.js (hunk 6 above)
Comment (Drapegnik) [reply pages/admin/preview.js:25]: to be determined)

**Pull request merged**